src/libre/re_comp.c

```diff
@@ -108,6 +109,8 @@ re_comp_ast(struct ast_re *ast,
 
 	if (!comp_iter(&env, x, y, ast->expr)) { goto error; }
 
```

Comment:
Would you want to remove states with no outgoing edges here, too?

The AST->NFA conversion sometimes leaves some of those, but I think they're only ever from the unused alternates for anchors. So it might be simple enough to keep track of whether they're used during the AST->NFA and know to remove them there instead.

Reply:
But then if those leave something else hanging, it will take another pass to remove those, ... and at that point we might be better off using reference counting and a zero-count table instead of mark and sweep.

Reply:
Hm that's interesting.

About the dead states: My comment here was specifically about the ones introduced from the AST's unused anchor/not-anchor states. For these, I _think_ the AST->NFA transformation has access to everything it needs to know in order to remove those without leaving anything else hanging. I'm picturing that as just keeping a boolean, saying which of the anchor/not-anchor states got used, but that might be over-simplified and miss-handle some situation I missed.

About reference counting in general: That'd be good for replacing `fsm_trim()`, which is expected to handle FSM passed in by a user.

Reply:
In other words, is there any way removing them _can_ leave something else hanging?

Reply:
For the dead-end start/end states? I don't believe so, I'm just trying to avoid making too much of a special case of them.

Reply:
Ok, I don't mind either way. I do think this is a viable option, and should be cheap. I don't know what it means for complexity of code in that area though.